Nextcloud-DLNA community container (#3614)
nextcloud/all-in-one: community-containers/dlna/readme.md

## DLNA server
This container bundles DLNA server for your Nextcloud files to be accessible by the clients in your local network. Simply run the container and look for a new media server `nextcloud-aio` in your local network.

### Notes

> - This is not working on Mac OS since it requires the `host` networking mode in docker, and it doesn't really share the host's network interfaces in this system

Reviewer: ```suggestion
- This is not working with Docker Desktop since it requires the `host` networking mode in docker, and it doesn't really share the host's network interfaces in this system
```

- If you have a firewall like ufw configured, you might need to open at least port 9999 TCP and 1900 UDP first in order to make it work.

> 

Reviewer: Should prpbably also add a note here that this is only for home networks and not for a vps for example...

Author: Yeah, you're right. I've added another note for that :)

Reviewer: I cannot see that note yet...

Also would be good if you would keep a note like this:
`See https://github.com/nextcloud/all-in-one/tree/main/community-containers#community-containers how to add it to the AIO stack`

Author: haha, that's because I forgot to push it :) there it goes. OK, I'll add a note you provided.

Reviewer: Nice!

## DLNA server

> This container bundles DLNA server for your Nextcloud files to be accessible by the clients in your local network. Simply run the container and look for a new media server Nextcloud-AIO in your local network.

Reviewer: Possibly?
```suggestion
This container bundles DLNA server for your Nextcloud files to be accessible by the clients in your local network. Simply run the container and look for a new media server `nextcloud-aio` in your local network.
```

> 

Reviewer: You will need to add some hints here, some ideas what should be written here are in https://github.com/nextcloud/all-in-one/tree/main/community-containers/plex#notes

Author: OK, how about now? I've also added a note about the existing nextcloud container requirement.

## DLNA server
This container bundles DLNA server for your Nextcloud files to be accessible by the clients in your local network. Simply run the container and look for a new media server `nextcloud-aio` in your local network.

### Notes

> - This container needs the `nextcloud-aio-nextcloud` container to be up and running to work

Reviewer: I would remove this notice. As I said in best case there should be a check maybe in the startup command of the container that checks if for example the given nextcloud domain reachable and only start up the service in that case?

Reviewer: You could use netcat-openbsd like this for example:
```bash
while ! nc -z "$NC_DOMAIN" 443; do
	echo "Waiting for Nextcloud to become online"
	sleep 5
done
```

Reviewer: In that case you would need to add a wrapper bash script as entrypoint and start the dlna server from there.

Reviewer: If a domain is given, it could run this check and wait for it to start. If not, it should skip this waiting...

Author: But I would have to put this script into the nextcloud-dlna container, right? I would rather not do that in the nextcloud-dlna container for two reasons:
* nextcloud-dlna doesn't need an access to the Nextcloud APP via HTTP, so checking it and waiting for it in most cases would make no sense (it only needs an access to the DB and `data` folder)
* Nextcloud is not always exposed on 443 port, there is a chance that it's a backend installation using port 80 (or other).

Reviewer: I've created a PR: https://github.com/thanek/nextcloud-dlna/pull/17. Let me know what you think :)

Author: Nice one! I'll take a look at it and let you know, looks pretty good at first glance :)

Author: Great. I've just released version 1.1.2 of nextcloud-dlna - it contains your wrapper and is aware of nextcloud presentce. Thanks!

Reviewer: nice :)

- This is not working on Docker Desktop since it needs `network_mode: host` in order to work correctly

> - This is not working on Mac OS since the `host` networking mode doesn't really share the host's network interfaces in this system

Reviewer: I think Docker Desktop includes macOS?

Author: Those two lines probably mean the same - the `host` mode on Mac OS is unusable for DLNA to work. I'll change that.

> - It is needed to allow TCP network traffic on port 9999 to the AIO instance and the UDP (port 1900) from it to the local network

Reviewer: BTW, now that I am reading this, don't you think using bridged networking and exposing 9999/tcp and 1900/udp should make it work?

> - It is needed to allow TCP network traffic on port 9999 to the AIO instance and the UDP (port 1900) from it to the local network

Author: Not sure, I'm not an expert in docker network settings. I've just glanced at plex and miniDLNA docker containers as an inspiration ;)
From what I know the requirements are:
* container must be able to make multicast requests to the local network on UDP, port 1900
* container must be able to receive the UDP requests from other machines in the local network
* in the above messages, DLNA messages contain the `PresentationURL` which should be a local network accessible address for TCP (HTTP) requests such as 192.168.0.10:9999.

> - It is needed to allow TCP network traffic on port 9999 to the AIO instance and the UDP (port 1900) from it to the local network

Reviewer: Have you already tried using
`"internal_port": "9999",`
and then adding 
```
            "ports": [
                {
                  "ip_binding": "",
                  "port_number": "9999",
                  "protocol": "tcp"
                },
                {
                    "ip_binding": "",
                    "port_number": "1900",
                    "protocol": "udp"
                }
            ],
```

> - It is needed to allow TCP network traffic on port 9999 to the AIO instance and the UDP (port 1900) from it to the local network

Author: I will check this out once I start my Linux box again :)

> - It is needed to allow TCP network traffic on port 9999 to the AIO instance and the UDP (port 1900) from it to the local network

Author: OK, I've  tested the `bridge` net mode and standard container net mode and they both failed. Even if the bridged network seems to use the same network interface as the host, since the container has a different address, the `PresentationURL` is unusable for other devices in the local network to reach the DLNA server with it. The container mode not only has a different address issue (just like `bridge`) but also it doesn't share the network interface so the container is unable to listen for responses from the local network devices at all. So, I guess the `host` option is the only way to go. Maybe it would be possible to create a network that would be accessible for both containers and other local network devices, but it doesn't seem to be simple (at least for me) and I suppose It would be easy to implement it in the AIO as a generic solution for every network configuration. But, as I said, I'm not an expert and maybe there is some better solution - I just don't know it.

> - It is needed to allow TCP network traffic on port 9999 to the AIO instance and the UDP (port 1900) from it to the local network

Reviewer: > . So, I guess the `host` option is the only way to go.

All right. Thanks for testing!
